packages/aws-cdk-lib/aws-sns-subscriptions/lib/subscription.ts

```diff
@@ -24,5 +28,60 @@ export interface SubscriptionProps {
    *
    * @default - No dead letter queue enabled.
    */
-  readonly deadLetterQueue?: IQueue;
+  readonly deadLetterQueue?: sqs.IQueue;
+}
+
+/**
+ * Generic subscription target
+ */
+export abstract class Subscription implements sns.ITopicSubscription {
+  constructor(private readonly subscriber: IResource) {
+  }
+
+  /**
+   * TODO Doc
+   *
+   * @see https://docs.aws.amazon.com/sns/latest/dg/sns-cross-region-delivery.html
+   *
+   * @param topic The topic to subscribe to
+   * @throws Error if the queue and target are in different regions and both of those regions are opt-in
+   * @throws Error if the queue and target are in different regions and if the subscriber is a Lambda function in an opt-in region
+   * @returns the grant principal
+   */
+  public generateGrantPrincipal(topic: sns.ITopic): iam.IPrincipal {
+    const [subscriberRegion, topicRegion] = [this.subscriber, topic].map(({ stack }) => stack.region);
+    // TODO check resolved?
+    if (subscriberRegion === topicRegion) {
+      return new iam.ServicePrincipal('sns.amazonaws.com');
+    };
+
+    const [isSubscriberRegionOptIn, isTopicRegionOptIn] = [subscriberRegion, topicRegion].map((region) => regionInfo.Fact.find(
+      region,
+      regionInfo.FactName.IS_OPT_IN_REGION,
+    ));
```

Comment:
Although all the [SNS supported opt-in Regions](https://docs.aws.amazon.com/sns/latest/dg/sns-cross-region-delivery.html) are `IS_OPT_IN_REGION`, not all `IS_OPT_IN_REGION` are supported:

https://github.com/aws/aws-cdk/blob/bf026bdd8557305d427510af49f1bc538d439cb6/packages/aws-cdk-lib/region-info/lib/aws-entities.ts#L52-L69

I'm not sure if this is something the CDK should maintain or let users "just figure it out" when their SNS messages don't show up. A happy medium could be to store the list of regions known to be supported, and only warn the user that the opt-in region they are using might be unsupported, and to open an issue if that's incorrect

Reply:
Turns out the region list in [SNS supported opt-in Regions](https://docs.aws.amazon.com/sns/latest/dg/sns-cross-region-delivery.html) is inaccurate, or at least not up to date. I was able to successfully run the integ tests in `ca-west-1`, which is an opt-in region not listed in the documentation table.

I'm going to assume that all opt-in regions are supported and this is just a case of stale documentation, but confirmation from the SNS team would be appreciated